src/v/kafka/security/scram_algorithm.cc

```diff
@@ -0,0 +1,205 @@
+#include "kafka/security/scram_algorithm.h"
+
+#include "utils/base64.h"
+#include "utils/to_string.h"
+#include "utils/utf8.h"
+
+#include <boost/algorithm/string.hpp>
+
+namespace kafka {
+
+// ALPHA / DIGIT / "/" / "+"
+// NOLINTNEXTLINE
+#define BASE64_CHAR "[a-zA-Z0-9/+]"
+
+// base64-4 = 4base64-char
+// base64-3 = 3base64-char "="
+// base64-2 = 2base64-char "=="
+// base64   = *base64-4 [base64-3 / base64-2]
+// NOLINTNEXTLINE
+#define BASE64                                                                 \
+    "(?:" BASE64_CHAR "{4})*(?:" BASE64_CHAR "{3}=|" BASE64_CHAR "{2}==)?"
+
+// Printable ASCII except ","
+// NOLINTNEXTLINE
+#define PRINTABLE "[\\x21-\\x2b\\x2d-\\x7e]+"
+
+// NOLINTNEXTLINE
+#define EXTENSIONS "(,[a-zA-Z]+=[\\x01-\\x2b\\x2d-\\x7f]+)*"
+
+// UTF8-char except NUL, "=", and ","
+// NOLINTNEXTLINE
+#define VALUE_SAFE_CHAR "[\\x01-\\x2b\\x2d-\\x3c\\x3e-\\x7f]"
+
+// 1*(value-safe-char / "=2C" / "=3D")
+// NOLINTNEXTLINE
+#define SASLNAME "(?:" VALUE_SAFE_CHAR "|=2C|=3D)+"
+
+// value-char = value-safe-char / "="
+// value = 1*value-char
+// NOLINTNEXTLINE
+#define VALUE "[\\x01-\\x2b\\x2d-\\x7f]+"
+
+// NOLINTNEXTLINE
+#define ALPHA "[a-zA-Z]"
+
+// clang-format off
+static constexpr const char* client_first_message_re =
+    "n,(?:a=(" SASLNAME "))?,(m=" VALUE ",)?n=(" SASLNAME "),"
+    "r=(" PRINTABLE ")(," ALPHA "+=" VALUE ")*";
+
+static constexpr const char* client_final_message_re =
+    "c=(" BASE64 "),r=(" PRINTABLE ")" EXTENSIONS ",p=(" BASE64 ")";
+// clang-format on
+
+std::ostream& operator<<(std::ostream& os, const scram_credential& cred) {
+    fmt::print(
+      os,
+      "salt {} server_key {} stored_key {} iterations {}",
+      cred._salt,
+      cred._server_key,
+      cred._stored_key,
+      cred._iterations);
+    return os;
+}
+
+client_first_message::client_first_message(bytes_view data) {
+    static const std::regex re(
+      client_first_message_re, std::regex::ECMAScript | std::regex::optimize);
+
+    // std::regex really likes using std::string
```

Comment:
[That can be fixed](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAM1QDsCBlZAQwBtMQBGAFlICsupVs1qhkAUgBMAISnTSAZ0ztkBPHUqZa6AMKpWAVwC2tEJNJb0AGTy1MAOWMAjTMRABmAAykADqgWE6rR6hiZmvv6BdDZ2jkYubl6KypiqQQwEzMQEIcam5koqanQZWQQxDs6uHt4Kmdm5YQX15baV8dVeAJSKqAbEyBwA5FLutsiGWADU4u46xJjAmAAes9jingCCo%2BOTmDNzdcS2wGsb25JjtBMG07M6RycA%2BgBueJgA7mdbO9d7Bzp1EdMMwjN9tlsDAFRFMFC8jMwCMgEAcACKwgjoEAgBFIhBPBYKAysAgKe51LEgR6iV7vD7YtC0OpPQiuREkM7uWSQ6HAWEvIlOJ645FojGUwXCxHI8mY7HU4C0z4MujM1nEdnETnciGbWysNpTJyofRTBZLZZSvEQCnygjHGlvT780jnKbuj2er0euEihBSABsUyMrq23vDXsZdXF2PNK0DU0woc2EdTMZAcctUcy9AU2L9TyowkVBAAnj59kXmMAFGK3WmG7aM4sVk9syJSfnpfisFRmMSCF0ZgB2HWehYEfq0dOZq3Im0vAB0LmAtggPX5i8s69Iwd3SamVZrXVmOvEw9R53OtgIweYa5PP1H9fdTczZswECkkggACkoYOECLgAVCekiSCeXIvumCqwkYmAKAo1aft%2B/51JsGQOqc4GQWOHoAPT4VM/ZEHeeIIXWYY%2BnKVJ%2BnB3antBb4tpafo2vBiHIbuCh%2BruCy4UxNFoAYt73Pc9F4uIACs0iSNJ6JiXM6aWKwjE/FRr40QqSofPyHFIUs7EIQZmACRpendhZklQeZs5sb6xlcVZyJ8aZakptRlLCaJcziTxDEyXJUkKb5SlNip7nnqiQw9KwIBDFJQykKYQyeElqDxTochyLCfQDPsoycElBDxWlXQ9AgIJYG4O4ANYgFJkiLsOACckieOBrXtZ1LVCPF3BJSlaWkBlQxJXm3glalMWkHAsAwIgKCoEYPh4OwZAUBAaArWt1QTCIwCcJwnjeFQa0EK4eYQE4pVJU4thZKW8VFaQ23wfQADytCsE902kFgCKiOwt3/XgCxpC8CEgysqQicMw03soIMGk4GrEKWehYM9xXHEY2M9DQ9BMGwHA8PwgjCKIKDZTIQh4E4eaQD0qA%2BMUTLxQAtBSsyohIMhyJww5TBzH2SMLFIozzKN87IMjjSkaQaBAliNKYnAWNoFRxAkgh%2BAEbOq7rkRs1rVRuOrhSpGzpQNPoeSCJbiu0DbrSxGbDstIbFstKbHTmz0Ch5YMXCxfFiXJSDo3LAAHAGHMBtwUz7TCR2Lp4adTBAuCECQMyXOrUx6Dt615%2B4nBDll/Ny6QU1lRVVXVHVDXeHFQwDaQeNSd4Q3pfF40gJNt09HNi19AQPgieQlDbat62CJg%2BBEObpCE4wLDA2THwaj4%2BN9Qlg2R/FhVTB8hAojHccJ0nlN8qn6fDbXMX18w1WUKHbdJZ33eH2NigDzXQ8ej1W4MORc7VWrDmHOAlqkDJDq1bu4A%2Bf1RrFUAXvSQSDhooIAdNcqpBIbEACBobgQA%3D%3D%3D)

Reply:
ahh, thanks! but we switched to ctre.